File: packages/python-sdk/strait/authoring/_dag_validation.py
"""DAG validation using Kahn's algorithm."""

from __future__ import annotations

from typing import TYPE_CHECKING

from strait._errors import DagValidationError

if TYPE_CHECKING:
    from strait.authoring._steps import Step


def validate_dag(steps: list[Step]) -> list[str]:
    """Validate a DAG of workflow steps using Kahn's algorithm.

    Returns the topologically sorted step refs, or raises DagValidationError.
    """
    if not steps:
        return []

    refs = [s.step_ref() for s in steps]

    _check_duplicate_refs(refs)

    ref_set = set(refs)
    _check_missing_refs(steps, ref_set)

    return _topological_sort(steps, refs)


def _check_duplicate_refs(refs: list[str]) -> None:
    seen: set[str] = set()
    duplicates: list[str] = []
    for ref in refs:
        if ref in seen:
            duplicates.append(ref)
        seen.add(ref)

    if duplicates:
        raise DagValidationError(
            f"Duplicate step refs: {', '.join(duplicates)}",
            duplicate_refs=duplicates,
        )


def _check_missing_refs(steps: list[Step], all_refs: set[str]) -> None:
    missing: list[str] = []
    for step in steps:
        for dep in step.base_options().depends_on:
            if dep not in all_refs:
                missing.append(dep)

    if missing:
        raise DagValidationError(
            f"References to non-existent steps: {', '.join(missing)}",
            missing_refs=missing,
        )


def _topological_sort(steps: list[Step], refs: list[str]) -> list[str]:
    in_degree: dict[str, int] = {ref: 0 for ref in refs}
    adjacency: dict[str, list[str]] = {ref: [] for ref in refs}

    for step in steps:
        for dep in step.base_options().depends_on:
            adjacency[dep].append(step.step_ref())
            in_degree[step.step_ref()] += 1

    queue = [ref for ref in refs if in_degree[ref] == 0]
    sorted_refs: list[str] = []

    while queue:
        node = queue.pop(0)
        sorted_refs.append(node)
        for neighbor in adjacency[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    if len(sorted_refs) != len(steps):
        sorted_set = set(sorted_refs)
        in_cycle = [ref for ref in refs if ref not in sorted_set]
        raise DagValidationError(
            f"Circular dependency detected involving steps: {', '.join(in_cycle)}",
            cycles=in_cycle,
        )

    return sorted_refs

File: packages/go-sdk/authoring/dag_validation.go
package authoring

import (
	"fmt"
	"strings"

	strait "github.com/strait-dev/go-sdk"
)

// ValidateDag validates a DAG of workflow steps using Kahn's algorithm.
// It returns the topologically sorted step refs or an error if the DAG is invalid.
func ValidateDag(steps []Step) ([]string, error) {
	if len(steps) == 0 {
		return nil, nil
	}

	refs := make([]string, len(steps))
	for i, s := range steps {
		refs[i] = s.StepRef()
	}

	if err := checkDuplicateRefs(refs); err != nil {
		return nil, err
	}

	refSet := make(map[string]bool, len(refs))
	for _, ref := range refs {
		refSet[ref] = true
	}

	if err := checkMissingRefs(steps, refSet); err != nil {
		return nil, err
	}

	return topologicalSort(steps, refs)
}

func checkDuplicateRefs(refs []string) error {
	seen := make(map[string]bool, len(refs))
	var duplicates []string

	for _, ref := range refs {
		if seen[ref] {
			duplicates = append(duplicates, ref)
		}
		seen[ref] = true
	}

	if len(duplicates) > 0 {
		return &strait.DagValidationError{
			Message:       fmt.Sprintf("Duplicate step refs: %s", strings.Join(duplicates, ", ")),
			DuplicateRefs: duplicates,
		}
	}
	return nil
}

func checkMissingRefs(steps []Step, allRefs map[string]bool) error {
	var missing []string

	for _, s := range steps {
		for _, dep := range s.BaseOptions().DependsOn {
			if !allRefs[dep] {
				missing = append(missing, dep)
			}
		}
	}

	if len(missing) > 0 {
		return &strait.DagValidationError{
			Message:     fmt.Sprintf("References to non-existent steps: %s", strings.Join(missing, ", ")),
			MissingRefs: missing,
		}
	}
	return nil
}

func topologicalSort(steps []Step, refs []string) ([]string, error) {
	inDegree := make(map[string]int, len(refs))
	adjacency := make(map[string][]string, len(refs))

	for _, ref := range refs {
		inDegree[ref] = 0
		adjacency[ref] = nil
	}

	for _, s := range steps {
		for _, dep := range s.BaseOptions().DependsOn {
			adjacency[dep] = append(adjacency[dep], s.StepRef())
			inDegree[s.StepRef()]++
		}
	}

	var queue []string
	for _, ref := range refs {
		if inDegree[ref] == 0 {
			queue = append(queue, ref)
		}
	}

	var sorted []string
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		sorted = append(sorted, node)

		for _, neighbor := range adjacency[node] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	if len(sorted) != len(steps) {
		var inCycle []string
		sortedSet := make(map[string]bool, len(sorted))
		for _, ref := range sorted {
			sortedSet[ref] = true
		}
		for _, ref := range refs {
			if !sortedSet[ref] {
				inCycle = append(inCycle, ref)
			}
		}
		return nil, &strait.DagValidationError{
			Message: fmt.Sprintf("Circular dependency detected involving steps: %s", strings.Join(inCycle, ", ")),
			Cycles:  inCycle,
		}
	}

	return sorted, nil
}
